plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Auction.hs

```diff
@@ -0,0 +1,269 @@
+--   off-chain use.
+machineClient
+    :: Scripts.ScriptInstance (StateMachine AuctionState AuctionInput)
+    -> AuctionParams
+    -> StateMachineClient AuctionState AuctionInput
+machineClient inst auctionParams =
+    let machine = auctionStateMachine auctionParams
+    in SM.mkStateMachineClient (StateMachineInstance machine inst)
+--BLOCK7
+
+type BuyerSchema = BlockchainActions .\/ Endpoint "bid" Ada
+type SellerSchema = BlockchainActions -- Don't need any endpoints: the contract runs automatically until the auction is finished.
+
+data AuctionLog =
+    AuctionStarted AuctionParams
+    | AuctionFailed SM.SMContractError
+    | CurrentState HighestBid
+    | BidSubmitted HighestBid
+    | AuctionEnded HighestBid
+    | CurrentStateNotFound
+    | TransitionFailed (SM.InvalidTransition AuctionState AuctionInput)
+    deriving stock (Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+
+-- | Client code for the seller
+auctionSeller :: Value -> Slot -> Contract SellerSchema SM.SMContractError ()
+auctionSeller value slot = do
+    self <- Ledger.pubKeyHash <$> ownPubKey
+    let params       = AuctionParams{apOwner = self, apAsset = value, apEndTime = slot }
+        inst         = scriptInstance params
+        client       = machineClient inst params
+        initialState = Ongoing HighestBid{highestBid = 0, highestBidder = self}
+
+    _ <- handleError
+            (\e -> do { logError (AuctionFailed e); throwError e})
+            (SM.runInitialise client initialState value)
+
+    logInfo $ AuctionStarted params
+    _ <- awaitSlot slot
+
+    r <- SM.runStep client Payout
+    case r of
+        SM.TransitionFailure i            -> logError (TransitionFailed i)
+        SM.TransitionSuccess (Finished h) -> logInfo $ AuctionEnded h
+        SM.TransitionSuccess s            -> logWarn ("Unexpected state after Payout transition: " <> show s)
+
+
+-- | Get the current state of the contract and log it.
+currentState :: StateMachineClient AuctionState AuctionInput -> Contract BuyerSchema SM.SMContractError (Maybe HighestBid)
+currentState client = SM.getOnChainState client >>= \case
+    Just ((TypedScriptTxOut{tyTxOutData=Ongoing s}, _), _) -> do
+        logInfo (CurrentState s)
+        pure (Just s)
+    _ -> do
+        logWarn CurrentStateNotFound
+        pure Nothing
+
+{- Note [Buyer client]
+
+In the buyer client we want to keep track of the on-chain state of the auction
+to give our user a chance to react if they are outbid by somebody else.
+
+At the same time we want to have the "bid" endpoint active for any bids of our
+own, and we want to stop the client when the auction is over.
+
+To achieve this, we have a loop where we wait for one of several events to
+happen and then deal with the event. The waiting is implemented in
+@waitForChange@ and the event handling is in @handleChange@.
+
+Updates to the user are provided via the log functionality. This is not as bad
```

Comment:
Maybe we should add a `notify` function stub, even if it does nothing right now except log?

Reply:
hm, I think it might be confusing to have notifications from `notify` mixed with regular log messages. At least here we are clear that what we are doing is a temporary workaroung.
I added a ticket for doing it properly (https://jira.iohk.io/browse/SCP-1837)

Reply:
What I meant is that we have a notify function that we can later make do actual notifications. As it is now, if we later add notification support we have to go and remember all the places that should be using notifications. If we have a function we just change it (or we have to change it a lot, and then we get compile errors at all the use sites to fix).

Reply:
Ok. I'll do that (in a separate PR)